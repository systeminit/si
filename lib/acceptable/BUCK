load("@prelude-si//:macros.bzl", "rust_library")

rust_library(
    name = "acceptable",
    features = [
        "default",
        "derive",
        "deserialize",
        "nats-headers",
        "serialize",
    ],
    deps = [
        "//lib/acceptable-macros:acceptable-macros",
        "//lib/nats-std:nats-std",
        "//third-party/rust:async-nats",
        "//third-party/rust:ciborium",
        "//third-party/rust:remain",
        "//third-party/rust:serde",
        "//third-party/rust:serde_json",
        "//third-party/rust:thiserror",
        "//third-party/rust:ulid",
    ],
    srcs = glob([
        "src/**/*.rs",
    ]),
)
