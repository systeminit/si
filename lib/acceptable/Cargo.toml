[package]
name = "acceptable"
version.workspace = true
authors.workspace = true
license.workspace = true
repository.workspace = true
edition.workspace = true
rust-version.workspace = true
publish.workspace = true

[features]
default = [
  "derive",
  "deserialize",
  "nats-headers",
  "serialize",
] # NOTE: if extracted, then default should probably be empty
derive = ["acceptable-macros"]
deserialize = ["ciborium", "serde_json"]
nats-headers = ["async-nats", "nats-std"]
serialize = ["ciborium", "serde_json"]

[dependencies]
acceptable-macros = { path = "../../lib/acceptable-macros", optional = true }
async-nats = { workspace = true, optional = true }
ciborium = { workspace = true, optional = true }
nats-std = { path = "../../lib/nats-std", optional = true }
remain = { workspace = true }
serde = { workspace = true }
serde_json = { workspace = true, optional = true }
thiserror = { workspace = true }
ulid = { workspace = true }
