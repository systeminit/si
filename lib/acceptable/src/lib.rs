//! # Acceptable
//!
//! Traits and behaviors that enable negotiable API types that understand their content type,
//! message type, and version.
//!
//! These message types are meant to be upgradable, meaning that older messages can be understood
//! and acted upon on newer services. Conversely older deployed services are able to recognize
//! incoming messages which are newer that what the service is able to support.
//!
//! If feature-enabled, these API types can determine their deserialization strategy based on
//! metadata external to the message bytes, typically via HTTP or NATS headers (a NATS-based
//! implementation is available via the `nats-headers` feature). Similarily if feature-enabled,
//! these types can also determine how to serialize their content to bytes for transmission.
//!
//! Finally another feature-enabled capability allows API authors to derive several trait
//! implementations, drastically cutting down on the amount of boilerplate when crafting request or
//! response types.

#![warn(
    bad_style,
    clippy::missing_panics_doc,
    clippy::panic,
    clippy::panic_in_result_fn,
    clippy::unwrap_in_result,
    clippy::unwrap_used,
    dead_code,
    improper_ctypes,
    missing_debug_implementations,
    missing_docs,
    no_mangle_generic_items,
    non_shorthand_field_patterns,
    overflowing_literals,
    path_statements,
    patterns_in_fns_without_body,
    unconditional_recursion,
    unreachable_pub,
    unused,
    unused_allocation,
    unused_comparisons,
    unused_parens,
    while_true
)]

pub(crate) mod all_versions;
pub(crate) mod container;
pub(crate) mod content_info;
pub(crate) mod error;
pub(crate) mod id;
#[cfg(feature = "deserialize")]
pub(crate) mod negotiate;
pub(crate) mod versioned;

#[cfg(feature = "nats-headers")]
pub use crate::error::HeaderMapParseMessageInfoError;
pub use crate::{
    all_versions::AllVersions,
    container::{
        Container,
        IntoContainer,
        SupportsContainers,
    },
    content_info::{
        ContentInfo,
        ContentType,
        MessageType,
        MessageVersion,
    },
    error::{
        BoxError,
        ContentInfoError,
        UnsupportedDefaultContentTypeError,
        UpgradeError,
    },
    id::RequestId,
    versioned::Versioned,
};
#[cfg(feature = "deserialize")]
pub use crate::{
    container::DeserializeContainer,
    error::{
        DeserializeError,
        NegotiateError,
    },
    negotiate::Negotiate,
};
#[cfg(feature = "serialize")]
pub use crate::{
    container::SerializeContainer,
    error::SerializeError,
};

const CONTENT_TYPE_CBOR: &str = "application/cbor";
const CONTENT_TYPE_JSON: &str = "application/json";

#[cfg(feature = "derive")]
pub use acceptable_macros::*;
