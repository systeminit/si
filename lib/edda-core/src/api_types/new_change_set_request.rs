use acceptable::{
    AllVersions,
    Container,
    CurrentContainer,
    IntoContainer,
    UpgradeError,
};
use serde::Deserialize;

mod v1;

pub use self::v1::NewChangeSetRequestV1;

#[remain::sorted]
#[derive(AllVersions, CurrentContainer, Clone, Debug, Deserialize, Eq, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum NewChangeSetRequestAllVersions {
    #[acceptable(current)]
    V1(NewChangeSetRequestV1),
}

impl IntoContainer for NewChangeSetRequestAllVersions {
    type Container = NewChangeSetRequest;

    fn into_container(self) -> Result<Self::Container, UpgradeError> {
        match self {
            Self::V1(inner) => Ok(Self::Container::new(inner)),
        }
    }
}

#[cfg(test)]
mod test {
    use std::{
        error::Error,
        fmt,
        fs::File,
        io::{
            self,
            BufRead as _,
            BufReader,
            Read as _,
        },
        path::Path,
    };

    use serde::{
        Serialize,
        de::DeserializeOwned,
    };

    /// Tests that a versioned object will always serialize to the same representation, no matter
    /// what future versions or changes to the object.
    ///
    /// NOTE: It is imperative that incrementatl refactorings do not lead to a version-incompatible
    /// or serialize-incompatible change. If a test fails and it's because there is a diff to the
    /// commiteed `.snap` file, this should be considered a failed refactoring. The remediation is
    /// to *not* NewChangeSet the `.snap` file but rather to fix the code so that the `.snap` format is
    /// 100% preserved.
    pub fn assert_serialize(name: &str, version: u64, serialize: impl Serialize) {
        insta::with_settings!({
            snapshot_path => format!("new_change_set_request/snapshots-v{version}"),
            prepend_module_to_snapshot => false,
            omit_expression => true,
            description => concat!(
                "\n",
                "\n",
                "!!!\n",
                "!!! System Initiative Developers:\n",
                "!!!\n",
                "!!! IMPORTANT:\n",
                "!!!\n",
                "!!!     The contents of this snapshot should *never* be modified as it\n",
                "!!!     represents the serialization of a versioned Rust type. If a tests fails\n",
                "!!!     with this warning, then something about a Rust type has changed\n",
                "!!!     the wire serialization of this type and would represent a potential\n",
                "!!!     production outage or data corruption.\n",
                "!!!\n",
                "!!!     Consider this an erroneous behavioral change of the Rust code and *not*\n",
                "!!!     an out-of-date snapshot or fixture!\n",
                "!!!\n",
                "\n",
                "\n",
            ),
        }, {
            insta::assert_json_snapshot!(name, serialize);
        });
    }

    pub fn assert_deserialize<T>(snapshot_name: &str, version: u64, expected: T)
    where
        T: fmt::Debug + DeserializeOwned + PartialEq,
    {
        let actual: T = read_from_snapshot(
            snapshot_name,
            &format!("new_change_set_request/snapshots-v{version}"),
        )
        .expect("failed to deserialize from snapshot");
        dbg!(&actual);
        assert_eq!(actual, expected);
    }

    fn read_from_snapshot<T>(name: &str, path: &str) -> Result<T, Box<dyn Error>>
    where
        T: fmt::Debug + DeserializeOwned + PartialEq,
    {
        let glob = format!("{path}/{name}.snap");

        let mut maybe_obj_result = None;
        insta::glob!(&glob, |path| {
            let bytes = read_snapshot_content(path).unwrap();

            maybe_obj_result = Some(serde_json::from_slice(&bytes).map_err(Into::into));
        });

        match maybe_obj_result {
            Some(object_result) => object_result,
            None => Err(Box::new(io::Error::other(format!(
                "snapshot not found: {glob}"
            )))),
        }
    }

    // Implementation is adapted from the [`insta::Snapshot::from_file`] function.
    //
    // See: <https://github.com/mitsuhiko/insta/blob/62bb0a3/insta/src/snapshot.rs#L339-L421>
    fn read_snapshot_content(path: &Path) -> Result<Vec<u8>, Box<dyn Error>> {
        let mut f = BufReader::new(File::open(path)?);

        // Skip through the snapshot metadata, that being the first YAML document, where YAML
        // documents are delimited with a `"---"` line
        {
            let mut buf = String::new();
            f.read_line(&mut buf)?;
            if buf.trim_end() == "---" {
                loop {
                    let read = f.read_line(&mut buf)?;
                    if read == 0 {
                        break;
                    }
                    if buf[buf.len() - read..].trim_end() == "---" {
                        break;
                    }
                }
            }
        }

        let mut buf = Vec::new();
        f.read_to_end(&mut buf)?;

        Ok(buf)
    }
}
