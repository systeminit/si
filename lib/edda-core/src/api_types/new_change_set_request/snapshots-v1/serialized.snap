---
source: lib/edda-core/src/api_types/new_change_set_request.rs
assertion_line: 168
description: "\n\n!!!\n!!! System Initiative Developers:\n!!!\n!!! IMPORTANT:\n!!!\n!!!     The contents of this snapshot should *never* be modified as it\n!!!     represents the serialization of a versioned Rust type. If a tests fails\n!!!     with this warning, then something about a Rust type has changed\n!!!     the wire serialization of this type and would represent a potential\n!!!     production outage or data corruption.\n!!!\n!!!     Consider this an erroneous behavioral change of the Rust code and *not*\n!!!     an out-of-date snapshot or fixture!\n!!!\n\n\n"
---
{
  "id": "01JQCVVDHXYX6S9YCV773R13MG",
  "baseChangeSetId": "01JQCVVDHXYX6S9YCV773R13MG",
  "newChangeSetId": "01JQCVVDHXYX6S9YCV773R13MG",
  "toSnapshotAddress": "8662433c81792076e05cd71dc10f683af3a4d86344f58500309325af8744c36b"
}